pandas/plotting/_matplotlib/core.py

```diff
@@ -736,9 +835,21 @@ def _get_ax_layer(cls, ax, primary=True):
         else:
             return getattr(ax, "right_ax", ax)
 
+    def _col_idx_to_axis_idx(self, col_idx: int):
+        """Return the index of the axis where the column at col_idx should be plotted"""
+        if isinstance(self.subplots, list):
+            # Subplots is a list: some columns will be grouped together in the same ax
+            for group_idx, group in enumerate(self.subplots):
+                if col_idx in group:
+                    return group_idx
+        else:
```

Comment:
Is this ``else`` clause required? I guess, that it may be dropped.

Reply:
I suppose, that once you added return type, ``mypy`` started complaining.
Inside the loop you return only in case ``if col_idx in group``, this is the reason for the ``mypy`` error.
I suggest that you modify it:
```
for group_idx, group in enumerate(self.subplots):
    if col_idx in group:
    return group_idx
raise KeyError(f"Failed to find {col_idx} in {self.subplots}")
```
Or something like this.

Reply:
I think mypy is wrong here. Because of the validation done earlier, there *must* be a valid column. Raising an exception might make mypy happy but it would never be hit (because of the validation), and we wouldn't be able to test against it. I changed the for loop into a `next()`, which seems to trick mypy into being reasonable again.